File: services/competitions-service/src/models/competition.py
from datetime import datetime
import uuid
from typing import Optional, TYPE_CHECKING
from sqlalchemy import String, Integer, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
import enum

from src.models.base import Base

if TYPE_CHECKING:
    from modality import ModalityModel
    from sport_ruleset import SportRulesetModel
    from stats import StatsRuleSetModel


class CompetitionStatus(str, enum.Enum):
    PENDING = "pending"
    STARTED = "started"
    FINISHED = "finished"

class CompetitionSystem(str, enum.Enum):
    POINTS = "points"
    ELIMINATION = "elimination"
    MIXED = "mixed"

class CompetitionPhase(str, enum.Enum):
    GROUPS = "groups"
    ELIMINATION = "elimination"

class CompetitionModel(Base):
    __tablename__ = "competitions"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    modality_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("modalities.id"))
    
    name: Mapped[str] = mapped_column(String(100))
    status: Mapped[CompetitionStatus] = mapped_column(String, default="PENDING")
    sport_ruleset_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True), ForeignKey("sport_rulesets.id"), nullable=True)

    start_date: Mapped[datetime]
    end_date: Mapped[datetime]
    
    # Configs
    system: Mapped[CompetitionSystem] = mapped_column(String, default="POINTS")
    min_members_per_team: Mapped[int] = mapped_column(Integer, default=5)
    max_members_per_team: Mapped[int] = mapped_column(Integer, default=20)
    
    teams_per_group: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    teams_qualified_per_group: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)

    image: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    current_phase: Mapped[Optional[CompetitionPhase]] = mapped_column(String, nullable=True)
    
    # Relacionamentos
    # Note que importamos classes dentro de strings ou TYPE_CHECKING para evitar ciclo
    modality: Mapped["ModalityModel"] = relationship(
        "ModalityModel", 
        back_populates="competitions"
    )
    
    sport_ruleset: Mapped[Optional["SportRulesetModel"]] = relationship(
        "SportRulesetModel", 
        back_populates="competitions"
    )
    
    stats_ruleset: Mapped[Optional["StatsRuleSetModel"]] = relationship(
        "StatsRuleSetModel",
        back_populates="competition",
        uselist=False
    )
    
    @property
    def organization_slug(self) -> Optional[str]:
        """Retorna o organization_slug da modalidade associada."""
        return self.modality.organization_slug if self.modality else None

File: services/frontend/types/competition.ts
export enum CompetitionStatus {
  PENDING = 'pending',
  STARTED = 'started',
  FINISHED = 'finished',
}

export enum CompetitionSystem {
  POINTS = 'points',
  ELIMINATION = 'elimination',
  MIXED = 'mixed',
}

export interface SportRuleset {
  id: string;
  name: string;
  organization_slug?: string | null;
  segment_type: string;
  segments_regular_number: number;
  overtime_segments: number;
  penalty_segments: number;
  has_break_segments: boolean;
}

export interface SportRulesetCreate {
  name: string;
  segment_type: string;
  segments_regular_number: number;
  overtime_segments: number;
  penalty_segments: number;
  has_break_segments: boolean;
}

export enum CompetitionPhase {
  GROUPS = 'groups',
  ELIMINATION = 'elimination',
}

export interface Competition {
  id: string;
  name: string;
  modality_id: string;
  organization_slug?: string;
  start_date: string;
  end_date: string;
  status: CompetitionStatus;
  system: CompetitionSystem;
  min_members_per_team: number;
  max_members_per_team: number;
  image?: string;
  teams_qualified_per_group?: number;
  teams_per_group?: number;
  sport_ruleset_id?: string;
  sport_ruleset?: SportRuleset;
  stats_ruleset?: {
    id: string;
    name: string;
    description?: string;
    stats_types: CompetitionStat[];
  } | null;
  current_phase?: CompetitionPhase;
}

export interface CompetitionCreate {
  name: string;
  modality_id: string;
  start_date: string;
  end_date: string;
  system: CompetitionSystem;
  min_members_per_team: number;
  max_members_per_team: number;
  image?: string;
  teams_qualified_per_group?: number;
  teams_per_group?: number;
  ruleset?: SportRulesetCreate;
  sport_ruleset_id?: string;
  stats_ruleset?: {
    name: string;
    description?: string;
    stats_types: any[];
  };
  stats_ruleset_id?: string;
}

export interface CompetitionUpdate {
  name?: string;
  start_date?: string;
  end_date?: string;
  status?: CompetitionStatus;
  min_members_per_team?: number;
  max_members_per_team?: number;
  system?: CompetitionSystem;
  sport_ruleset_id?: string;
  stats_ruleset_mode?: "keep" | "none" | "new";
}

export interface GenerateStructureRequest {
  organization_id: string;
}

export interface GenerateStructureResponse {
  message: string;
  system: string;
  matches_created: number;
  lives_created: number;
  lives: any[];
}

export interface CompetitionStat {
  id: string;
  competition_id: string;
  name: string;
  abbreviation: string;
  description?: string;
  display_order: number;
}

export interface CompetitionStatCreate {
  name: string;
  abbreviation: string;
  description?: string;
  display_order?: number;
}

export interface Player {
  id: string;
  keycloak_id: string;
  team_id: string;
}

export interface TeamWithPlayers {
  id: string;
  name: string;
  abbreviation: string;
  logo_url?: string | null;
  /** ID do time no auth-service; preferir em links para /clubes/[id] */
  auth_team_id?: string | null;
  players: Player[];
}
